src/extension.ts

```diff
@@ -1,83 +1,134 @@
-};
+import { search } from './utils/search';
+import { matchSearchPhrase } from './utils/matchSearchPhrase';
+import { getConfig } from './config';
+import Parsers from './utils/parsers/index';
+import ParserAbstract from './utils/parsers/ParserAbstract';
+import *  as fs from 'fs';
+import *  as path from 'path';
+import * as os from 'os';
+import fetch from 'node-fetch';
+
+const tempDirectory = fs.mkdtempSync(path.join(os.tmpdir(), 'captainstack-'));
+let externalParsers: Promise<ParserAbstract>[] = getConfig().settings.externalParsers.filter((e) => e !== '').map(async (e) => {
+    const contents = await fetch(e).then((res) => res.text());
+    const name = e.split('.')[e.split('.').length - 2];
+    const filePath = path.join(tempDirectory, name) + '.ts';
+
+    fs.writeFileSync(filePath, contents);
+    return await import(filePath);
+});
+if (externalParsers.length !== 0) vscode.window.showInformationMessage('CaptainStack: Loaded ' + externalParsers.length + ' external parsers.');
+
+interface ExtendedStatusBarItem extends vscode.StatusBarItem {
+    idle?: () => void,
+    working?: (tooltip: string) => void
+}
 
 export function activate(_: vscode.ExtensionContext) {
+    const statusBar: ExtendedStatusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
+    statusBar.idle = (() => {
+        statusBar.text = 'Captain Stack $(gear)';
+        statusBar.tooltip = 'Idle';
+    });
+    statusBar.working = ((tooltip: string = 'Working...') => {
+        statusBar.text = 'Captain Stack $(gear~spin)';
+        statusBar.tooltip = tooltip;
+    })
+
+    statusBar.color = "pink";
+    statusBar.show();
+    if(statusBar.idle) statusBar.idle();
+
 
     const provider: vscode.CompletionItemProvider = {
         // @ts-ignore
-        provideInlineCompletionItems: async (document : vscode.TextDocument, position, context, token) => {
+        provideInlineCompletionItems: async (document: vscode.TextDocument, position, context, token) => {
 
-            const textBeforeCursor = 'Language: ' + lang.languages[document.fileName.split('.').pop() || ''] + '\n\n' + document.getText(
-                new vscode.Range(position.with(0, 0), position)
+            const textBeforeCursor = document.getText(
+                new vscode.Range(position.with(undefined, 0), position)
             );
 
-            if(run) {
-                await new Promise((resolve, reject) => {
-                    setInterval(() => {
-                        if(!run) resolve(null);
-                    }, 10)
-                })
+            const match = matchSearchPhrase(textBeforeCursor);
+            let items: any[] = [];
 
-                run = false;
-            }
+            if (match) {
+                if (statusBar.working) statusBar.working('Fetching snippets...');
+                let rs;
+                try {
+                    rs = await search(match.searchPhrase);
+                    if (rs) {
```

Comment:
i've seen you use a lot of nesting throughout the whole project. We prefer Guard-Clausing. This is a perfect case of where this could've been put to use. Please refactor.

Reply:
Sorry, I don't quite understand what I should do. Could you explain a bit more?

Reply:
Google Guard Clauses